### ISSUE TITLE
Vault init support for subkeys

### ISSUE THREAD

**Issue by Toady00** [hashicorp/vault#2372]
Original google group discussion [here](https://groups.google.com/forum/#!topic/vault-tool/_LUM5YYnui8)

Using vault 0.6.4. I have an offline master key, and a yubikey with 3 subkeys (encrypt, sign, auth). I also use the same keys on keybase. (Keybase doesn't not have my private master key). When trying to initialize vault with my keybase user, I get the following error.

```
$ vault init --key-shares 1 --key-threshold 1 --pgp-keys "keybase:toady00" --address "http://127.0.0.1:8200"
Error initializing Vault: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/sys/init
Code: 400. Errors:

* Error setting up encryption for PGP message: openpgp: invalid argument: cannot encrypt a message to key id ee49708f2def1f39 because it has no encryption keys
```

I haven't been able to find a way to export a subkey with openGPG, so if I export my gpg key and try using it directly I get the same error due to the fact that my master public key is only allowed to sign other keys.

From @jefferai replies to the google group discussion, vault is pulling the first key it finds from keybase, and in this case that isn't the correct key to use.

**Comment by jefferai** [comment]
My earlier assertion is not correct; after diving into the openpgp code, it seems like it does actually iterate through all subkeys looking for an encryption key (https://github.com/keybase/go-crypto/blob/master/openpgp/keys.go#L109)

Still looking...

**Comment by jefferai** [comment]
Putting some thoughts here:

* What needs to happen is that we need to iterate through the parsed entitylist and find an encryption key.
* There may be more than one, which potentially means adding support for specifying (via fingerprint and/or key id) which key one wants to use for encryption.
* Either way, need to iterate through the entity list to check each entity and then each entity's subkeys to find key ids to then check with the entity list to then verify whether that key id can be used for encryption to then get a key which leads back to an entity which can then be serialized to pass to the Vault API, except that in that case if the entity has more than one subkey we're in the same boat.
* So, properly fixing this probably also means adding support in the Vault API to be given a key id or fingerprint selector and having the server also be able to filter through. Have I mentioned how PGP sucks. It would be 10x easier to just say "gimme a serialized RSA public key thankyoumuch".

**Comment by bgupta** [comment]
@jefferai I'm having basically the same issue. After digging in I think this is the cause and resolution. https://github.com/keybase/go-crypto/pull/50

Looks like you pulled in the new fixed dependency on Mar 30: https://github.com/hashicorp/vault/commit/790261277397502e1565cac64abaea3ab6cbec0a#diff-851cdaf90f5ed1eb0c57d1e2fa2ac379

Guessing this will be resolved when you cut v0.7.1. I'd be willing to test and see if it resolves my issue, if you can point me to a nightly build.

**Comment by arthurlogilab** [comment]
I'm getting this bug (or a similar bug) with version v0.7.3, so I think (if the go-crypto fix has been integrated) that it hasn't fixed it.

**Comment by jefferai** [comment]
@arthurlogilab That go-crypto fix is might fix one issue, but as I mentioned above Vault doesn't have a way to select a particular subkey for encryption.

**Comment by horsehay** [comment]
I'm not using subkeys and the rekey is still failing with 0.7.3 because it's still combining two fingerprints.

**Comment by jefferai** [comment]
> I'm not using subkeys and the rekey is still failing with 0.7.3 because it's still combining two fingerprints.

What do you mean by "still combining two fingerprints"? This sounds like both something I've not heard of before and a different issue than this one.

**Comment by bgupta** [comment]
@jefferai For the record, the key that I was seeing this issue with stopped showing the issue after I upgraded to 0.7.2.

**Comment by jefferai** [comment]
Oh, nice!

**Comment by trinitronx** [comment]
Apparently this is happening also in the following case:

- User has multiple PGP keys on keybase.io
- First PGP key is old & expired or has no active encryption subkey
- Second PGP key is valid & has all subkey types

I'm surprised that Vault has no way to specify which PGP key to use for keybase users owning multiple PGP keys.  Keybase knows how to handle this because it just uses GPG key ID to specify which.  Of course multiple PGP keys will usually have different User IDs & emails attached to them.

**Comment by Toady00** [comment]
Has there been any movement on this lately? I'd offer to help but I'm horrible at golang. I just deployed vault and consul clusters at a Fortune 100 company using the registry.terraform.io modules. I'm waiting for the networking team to create a route to the clusters to verify they're up and running correctly, but I was really hoping that I'd be able to use my keys with Vault.

I opened this issue in February. Is there anything else I can help with short of actually doing the work?

**Comment by jefferai** [comment]
Hi @Toady00 ,

No movement on it lately. The fastest way to get it addressed would definitely be to take on the work yourself, or hope that we find time to get to it during the 0.10 dev cycle, or (if you're an Enterprise customer) reach out through your sales contact.

On our end, we have found since we first introduced the PGP capability way way back that a large number of users have no PGP infrastructure (and in many cases don't want to set one up, either because they think they will lose the keys or they don't really understand how to use it). We developed, but have not finished implementing, a new method of handling unseal keys that provides nearly all of the same benefits as using PGP keys but uses only Vault primitives and capabilities. So when we have time to tackle this problem, that's where we're inclined to spend it since it can have a much broader impact, which is a large part of the reason that we haven't put effort into this enhancement so far.

**Comment by trinitronx** [comment]
@jefferai :  It was my understanding that a minimum viable product here would be to use well known and trusted public gpg keyservers (e.g.: `hkps://hkps.pool.sks-keyservers.net`, `keys.gnupg.net`, `pgp.mit.edu`), or [keybase.io][1] itself to store the keys (if that's what you mean by "PGP infrastructure"?)

The key ids and/or fingerprints would be directly from GPG / PGP as is shown in the `keybase pgp list` command (or `gpg2 --list-keys --with-subkey-fingerprints --with-key-origin 'John Doe'`).  Anyone using GnuPG or PGP long enough would know what a key id or fingerprint is, and this feature is really for the usually more advanced user who has multiple keys.  We ended up running into this issue because I had put 2 GPG keys on [keybase.io][1], and were trying to pass these keybase users to `vault init`.

It was something like the original example:

```
vault init -key-shares=4 -key-threshold=2  \
 -pgp-keys="keybase:exampleuser1,keybase:exampleuser2,keybase:exampleuser3,keybase:exampleuser4"
```

Where:
 - one or more users have multiple keys
 - at least 1 key is expired and listed as first one for the user

This results in vault picking the first expired key for that user and giving the error like:

```
* Error setting up encryption for PGP message: openpgp: invalid argument: cannot encrypt a message to key id 123456789abcdef0 because it has no encryption keys
```

[1]: https://keybase.io/

**Comment by trinitronx** [comment]
Posting back some helpful information for those still seeing this issue.

After my team revisited this problem, we found that part of the issue causing this `vault` error was due to the first `keybase` key being used, but **this first key was expired**.  Vault was picking the very first key it saw as the encryption key, but since this key had expired the error was returned:

```
* Error setting up encryption for PGP message: openpgp: invalid argument: cannot encrypt a message to key id 123456789abcdef0 because it has no encryption keys
```

GPG would also return an error when trying to use this expired key.

To try and resolve this issue, we modified the key's expiry date to be in the future.  At first, this did not appear to work because only the expiry date on the DSA key was the one which expiry was modified.  After we found this issue keybase/keybase-issues#2072, a [comment there pointed out instructions to modify the ElGamal encryption **sub-key also**][1].  After this, `vault init` worked!

Additionally, this was never the PGP key that was intended to be used (since it was expired, it should have remained so).  So, we tried using `keybase pgp drop` to remove the expired key, then re-add it back again as the second key in the list.  This worked!  So, the following process might work if anyone is having a similar issue with first key being an old expired one (or just not the one you wish to encrypt your Vault seal tokens with).

```
keybase pgp list
# Find the "Keybase Key ID:" for the one you wish to remove from keybase, use as $KEYBASE_KEY_ID below
keybase pgp drop $KEYBASE_KEY_ID

# Now if you want to re-add that old key as *last* in the list (just for historical or keybase proof purposes)
gpg2 --list-secret-keys
# Find the full GPG key id in gpg2 output above, and use as $GPG_KEY_ID below
gpg2 --export-secret-keys $GPG_KEY_ID | keybase pgp import
```

In summary, the issue was caused by `vault init` using the **first key** listed from `keybase pgp list`, which was expired and had an expired sub-key.  Changing the order of the keys is possible via `keybase pgp drop` and re-import.

[1]: https://github.com/keybase/keybase-issues/issues/2072#issuecomment-183731955

**Comment by gnarea** [comment]
I experienced this issue because the key had recently expired. I didn't have multiple subkeys.

**Comment by greut** [comment]
Using another fork of `x/crypto` seems good enough to address this issue, see #16224
